File: algorithms/2_medium/M/02064_minimized_maximum_of_products_distributed_to_any_store/binary_search.c
inline static int groups_state(const int *restrict arr, const int len, const int val, const int target) {
    int cnt = 0;
    for (int i = 0; i < len; i += 1) {
        cnt +=  (arr[i] + val - 1) / val;
        if (cnt > target) {
            return 1;
        }
    }

    return cnt == len ? 0 : -1;
}

int minimizedMaximum(const int n, const int *restrict quantities, const int quantities_len) {
    assert(n >= quantities_len && quantities_len > 0);

    int lo = 1;
    int hi = INT_MAX;

    while (lo <= hi) {
        const int mid = lo + (hi - lo) / 2;
        switch (groups_state(quantities, quantities_len, mid, n)) {
            case 1:
                lo = mid + 1;
                break;
            default:
                hi = mid - 1;
                break;
        }
    }

    return lo;
}

File: algorithms/2_medium/M/02064_minimized_maximum_of_products_distributed_to_any_store/binary_search.rs
use std::cmp::Ordering;

impl Solution {
    #[inline(always)]
    fn groups_state(arr: &[i32], val: i32, target: i32) -> Ordering {
        let mut cnt = 0_i32;
        for &a in arr {
            cnt +=  (a + val - 1) / val;
            if cnt > target {
                return Ordering::Greater;
            }
        }

        if cnt == target {
            return Ordering::Equal;
        }

        Ordering::Less
    }

    pub fn minimized_maximum(n: i32, quantities: Vec<i32>) -> i32 {
        assert!((n as usize) >= quantities.len() && !quantities.is_empty());

        let max_num = *quantities.iter().max().unwrap();
        let mut lo: i32 = 1;
        let mut hi: i32 = max_num;

        while lo <= hi {
            let mid = lo + (hi - lo) / 2;
            match Self::groups_state(&quantities, mid, n) {
                Ordering::Greater => lo = mid + 1,
                _ => hi = mid - 1,
            }
        }

        lo
    }
}
